mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala

```diff
@@ -78,13 +76,34 @@ private[spark] trait DecisionTreeModel {
 
   /** Convert to spark.mllib DecisionTreeModel (losing some information) */
   private[spark] def toOld: OldDecisionTreeModel
+
+  /**
+   * @return an iterator that traverses (DFS, left to right) the leaves
+   *         in the subtree of this node.
+   */
+  private def leafIterator(node: Node): Iterator[LeafNode] = {
+    node match {
+      case l: LeafNode => Iterator.single(l)
+      case n: InternalNode =>
+        leafIterator(n.leftChild) ++ leafIterator(n.rightChild)
+    }
+  }
+
+  @transient private lazy val leafIndices: Map[LeafNode, Int] = {
+    leafIterator(rootNode).zipWithIndex.toMap
+  }
+
+  /**
+   * @return the index of leaf given a input vector.
+   *         The leave are indexed from zero by pre-order.
```

Comment:
Nit: leave -> leaves. I might just write: `@return the index of the leaf corresponding to the feature vector. Leaves are indexed in pre-order from 0.` (Same for other occurrences)